Support git-bash in Windows.

ipcjs opened the pull request.

ipcjs committed:
Support git-bash in Windows.

ipcjs committed:
add properties in package.json

jonsmithers commented:
This looks really cool! I will look at this tomorrow evening.

jonsmithers committed:
Obtain working directory in platform-agnostic way

(Windows will not have $HOME defined)

jonsmithers committed:
npm install --save-dev tslint

jonsmithers committed:
Fix working directory bug in windows and show friendlier error for users missing Git/bin/bash.exe

* Re-format filepath passed to unix "cd" command. Windows' different
  filepath format was causing the "cd" command to fail.
* Introduce an error dialog instructing Windows user to install Git when
  Git/bin/bash.exe is missing.

jonsmithers committed:
Show error message when using Gvim on Windows

jonsmithers commented on package.json:
```diff
@@ -54,6 +54,11 @@
+                    "type": "string",
+                    "default": null,
+                    "definition": "Shell path of integrated terminal."
+                },
```
I wonder if this should be removed. Why would a user want to change `integratedShellPath`?

ipcjs replied:
On windows, we can install git to other directories.😅
![ScreenClip](https://user-images.githubusercontent.com/4396864/68075716-6cbad600-fde6-11e9-8e08-95c3bdaf0468.png)

jonsmithers replied:
That makes sense. Good point :smile:

jonsmithers commented:
I noticed that there are actually two `bash.exe`s on my system: 
* `C:\Program Files\Git\bin\bash.exe`
* `C:\Windows\System32\bash.exe`.

`Git\bin\bash.exe` sees the `C:\` drive at `/c/`.
`System32\bash.exe` sees the `C:\` drive at `/mnt/c/`. 

It seems like `System32\bash.exe` is for a WSL I installed (Windows Subsystem for Linux).

I think `C:\Program Files\Git\bin\bash.exe` is more likely to exist on users' machines, but right now I'm seeing strange errors when running vim from Git\bin\bash.exe inside vscode. I don't know what changed!

```
--- Autocommands ---
Error detected while processing /etc/vimrc:
line   43:
E10: \ should be followed by /, ? or &
line   44:
E10: \ should be followed by /, ? or &
line   45:
E10: \ should be followed by /, ? or &
line   46:
E10: \ should be followed by /, ? or &
line   47:
E10: \ should be followed by /, ? or &
line   48:
E10: \ should be followed by /, ? or &
line   53:
E10: \ should be followed by /, ? or &
line   54:
E10: \ should be followed by /, ? or &
Error detected while processing function plug#end:
line   86:
E484: Can't open file C:Program Files (x86)Vimvim81/syntax/syntax.vim
Error detected while processing /c/Users/JonSm/_vimrc:
line  981:
E484: Can't open file C:Program Files (x86)Vimvim81/defaults.vim
Press ENTER or type command to continue
```

`system32\bash.exe` works fine for me (after fixing filepaths to use `/mnt/c` instead of `/c`).

jonsmithers commented:
I restarted vscode and those errors stopped. ¯\\\_(ツ)\_/¯

jonsmithers committed:
Refactor filepath re-formatting on windows

* Use the correct case for the /c/ directory.
* Move "is this windows?" check into function to reduce repeated code.
* Add path reformatting to tmp file returned by openArgsToScriptFile()

jonsmithers committed:
Add documentation for new option and rename it to "integrated-terminal.pathToShell"

Move default shellPath logic into getConfiguration();

The new name follows the existing naming convention for options.

ipcjs commented:
Maybe we should also support Windows Subsystem for Linux?

jonsmithers committed:
Fix tslint and reformat some code

jonsmithers commented:
It's a little difficult to support both. It seems like we'd have to launch a separate shell instance just to check where the `C:` drive is (in order to conditionally update _all_ paths to have a `/mnt` prefix).

```bash
test -d /mnt/c
exit $?
```

I also encountered some colorscheme issues when using WSL, but I didn't look into why.

For now, maybe we should only support git bash.exe. Maybe this will avoid unnecessary complexity.
